EDU-17144: Creating a new section in the CMS (#2637)
vtexdocs/dev-portal-content: docs/guides/CMS/Integrations/FastStore/extending-a-component.md

---

> title: "Extending a component"  

Reviewer: @mariana-caetano fiquei na dúvida do intuito desta documentação. Entendi pelo passo a passo do guia que é uma seção nova, do zero, não uma extensão de uma seção existente. Ai pra mim faz sentido mudar o título e também as linhas 11 e 13 para focar no caso de uso de criar uma seção nova (como é sua correspondente do hCMS https://developers.vtex.com/docs/guides/faststore/developing-and-overriding-components-creating-a-new-section)

Sobre extender uma seção existente, vejo que seria mais similar o caso de section overrides (https://developers.vtex.com/docs/guides/faststore/overrides-overview)
Como vocês vêem? @mariana-caetano @hellofanny

Reviewer 2: Concordo com Manu! Acho que discutimos isso rapidamente sobre a confusão está na nomeclatura (extending...)..
Mas esse guia realmente seria a mesma ideia do `Creating a new section` que temos hoje, mas para o CMS novo.

Author: Entendi, pessoal! Vou ajustar o título. Sobre essa parte de extensão de componentes, depois vejo com vocês como é feita no novo CMS.

> title: "Extending a component"  

Author: ```suggestion
title: "Creating a new section in the CMS"
```

hidden: false  
slug: "cms-extending-a-component"  
createdAt: "2026-02-19T12:50:00.813Z"  
updatedAt: "2026-04-30T11:48:00.813Z"  
---

> > ⚠️ This guide is valid only for [FastStore stores running v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version) or higher and using the [CMS](https://developers.vtex.com/docs/guides/cms-for-faststore-storefronts). For stores using a FastStore version below v4, see [this guide](https://developers.vtex.com/docs/guides/faststore/developing-and-overriding-components-creating-a-new-section).

Reviewer: "For stores using a FastStore version below v4, see [this guide]" ->
Trocaria para "For stores using hCMS, see [this guide]". Faz sentido para você, @hellofanny?

Reviewer 2: eita, melhor deixa bem explicito. 
Esse guia é pra stores que estão usando content platform (new CMS) seja na v3 ou v4 (acredito que aqui importa menos a versão da FS e mais o content source)

> > ⚠️ This guide is valid only for [FastStore stores running v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version) or higher and using the [CMS](https://developers.vtex.com/docs/guides/cms-for-faststore-storefronts). For stores using a FastStore version below v4, see [this guide](https://developers.vtex.com/docs/guides/faststore/developing-and-overriding-components-creating-a-new-section).

Author: ```suggestion
> ⚠️ This guide applies only to stores using the [CMS](https://developers.vtex.com/docs/guides/cms-for-faststore-storefronts) with FastStore versions 3 or 4. For stores using Headless CMS, please refer to [this guide](https://developers.vtex.com/docs/guides/faststore/developing-and-overriding-components-creating-a-new-section).
```

To reuse an existing layout or behavior (such as a banner, hero, or product shelf), you can extend a component and adapt it to your store brand or business rules without rewriting everything from scratch. By extending rather than creating components from scratch, you maintain visual and technical consistency and reduce maintenance.

In this guide, you will learn how to extend the **CallToAction** section in your FastStore project.

![call-to-action](https://vtexhelp.vtexassets.com/assets/docs/src/call-to-cation-admin___39afe92cbcd5c59606b7fa1db4ceb492.png)

> > ℹ️ In FastStore and CMS, a **section** is a type of component that contains other components and acts as a dynamic container in page layouts. For more details, see [Understanding Components and Sections](https://developers.vtex.com/docs/guides/understanding-components-and-sections).

Author: Link will work after the merge of #2638

> ⚠️ The CMS is currently available for stores running Faststore v3 and [FastStore v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version). If your store uses the [Headless CMS (legacy)](https://developers.vtex.com/docs/guides/faststore/headless-cms-overview), see the [Upgrading from Headless CMS (legacy) to CMS](https://developers.vtex.com/docs/guides/upgrading-from-headless-cms-legacy-to-cms-overview) track first.

### Install the Content plugin

After installing the CMS into your store account, install the Content plugin on your machine. For more information, see the [Content plugin](https://developers.vtex.com/docs/guides/content-plugin) guide.
</Steps>

## Instructions

### Step 1 - Creating the component

1. Open your store project in a code editor.
2. In the folder `src/components`, create the `CallToAction.tsx` file.
3. Add the following code to the `CallToAction.tsx` file:

   ```tsx
   import React from "react";  

> 

Reviewer 3: Add the $componentKey to match the component to the schema.
```suggestion

CallToAction.$componentKey = 'CallToAction';
```

6. In the `cms_component__CallToAction.jsonc` file, add the following:

   ```jsonc
   {  
     "$extends": [  
       "#/$defs/base-component"  
     ],  
     "$componentKey": "CallToAction",  
     "$componentTitle": "Call To Action",  
     "title": "Call To Action",  
     "description": "Get your 20% off on the first purchase!",  
     "type": "object",  
     "required": [  
       "title",  
       "link"  
     ],  
     "properties": {  
       "title": {  
         "title": "Title",  
         "type": "string"  
       },  
       "link": {  
         "title": "Link Path",  
         "type": "object",  
         "required": [  
           "text",  
           "url"  
         ],  
         "properties": {  
           "text": {  
             "title": "Text",  
             "type": "string"  
           },  
           "url": {  
             "title": "URL",  
             "type": "string"  
           }  
         }  
       }  
     }  
   }  
   ```

7. Generate the final schema to add the new section. To do this, run the following command in a terminal:

    ```bash

>     vtex content generate-schema cms/faststore/components cms/faststore/pages -o cms/faststore/schema.json -b vtex.faststore  

Reviewer 3: Default base is vtex.faststore, so let's remove it to avoid confusion. The content plugin detects the faststore version.

```suggestion
    vtex content generate-schema cms/faststore/components cms/faststore/pages -o cms/faststore/schema.json
```

1. In the project root, start the FastStore development server:

    ```bash
    yarn dev
    ```

2. Open the development server in your browser and check the terminal output for compilation errors related to `CallToAction`.

> ℹ️ The section won't render visually yet because it hasn't been added to any page in the CMS. This is done in [Step 4](#step-4---add-the-component-to-the-cms). At this point, you're only verifying that the development server starts and that your component code, schema, and registration are valid.

If the development server starts successfully and reports no compilation errors, you're ready to sync your changes with the CMS.

### Step 3 - Sync the changes with the CMS

Push the changes to sync with the CMS. Before doing it, make sure you're logged in to your VTEX account by running in the terminal the `vtex login {accountName}` command. Change the `{accountName}` to your store account, for example, `vtex login mystore`.

Push the schema generated to the CMS to reflect in the Admin. To do this, run the following command:

```shell

> vtex content upload-schema cms/faststore/schema.json

Author: ```suggestion
vtex content upload-schema -s faststore cms/faststore/schema.json
```

> vtex content upload-schema cms/faststore/schema.json

Reviewer 3: optional:
There is a new `-s {storeId}` param to avoid asking for the store name when running this upload-schema.

Using this new parameter, the user is not asked for the ID, which avoids confusion.
`vtex content upload-schema -s faststore cms/faststore/schema.json`

Author: Thanks @mateuspontes! I think it's worth mentioning, so I've made some changes to reflect your suggestions.

---
title: "Extending a component"  
hidden: false  
slug: "cms-extending-a-component"  
createdAt: "2026-02-19T12:50:00.813Z"  
updatedAt: "2026-04-30T11:48:00.813Z"  
---

> ⚠️ This guide is valid only for [FastStore stores running v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version) or higher and using the [CMS](https://developers.vtex.com/docs/guides/cms-for-faststore-storefronts). For stores using a FastStore version below v4, refer to [this guide](https://developers.vtex.com/docs/guides/faststore/developing-and-overriding-components-creating-a-new-section).

To reuse an existing layout or behavior (such as a banner, hero, or product shelf), you can extend a component and adapt it to your store brand or business rules without rewriting everything from scratch. By extending rather than creating components from scratch, you maintain visual and technical consistency and reduce maintenance.

> In this guide, you will learn how to extend the **CallToAction** section in your FastStore project.

Reviewer 4: ```suggestion
In this guide, you'll learn how to extend the **CallToAction** section in your FastStore project.
```

![call-to-action](https://vtexhelp.vtexassets.com/assets/docs/src/call-to-cation___e74ef106c68c9e8d1e99edb2e9840d7b.png)

> > ℹ️ In FastStore and CMS, a **section** is a type of component that contains other components and acts as a dynamic container in page layouts. For more details, see [Understanding Components and Sections](https://developers.vtex.com/docs/guides/understanding-components-and-sections).

Author: ```suggestion
> ℹ️ A **section** is a type of component that contains other components and acts as a dynamic container in page layouts. For more details, see [Understanding Components and Sections](https://developers.vtex.com/docs/guides/understanding-components-and-sections).
```

## Before you begin

<Steps>

> ### Install the CMS

Reviewer 5: @mariana-caetano I think this is duplicating details that we should keep in a separate Install CMS doc - instead of embedding in this document.

Author: Makes sense, @jamesthethird. I'll move this section's content to the Installation draft.

> ### Install the CMS

Author: 🎗️ @mariana-caetano: Ensure part of this content section is moved to the Installation draft.

Before extending a component, make sure the [CMS](https://developers.vtex.com/docs/guides/cms-for-faststore-storefronts) is enabled in your VTEX account. The CMS provides the Admin interface to configure and publish the **CallToAction** section you build in this guide.

To check whether the CMS is available in your account, log in to your VTEX Admin and go to **Storefront > Content**.

- If the **Content** option appears, the CMS is already enabled and you can proceed to the [next section](#install-the-content-plugin).

> - If the **Content** option is not available, [open a ticket with VTEX Support](https://help.vtex.com/en/support) requesting the CMS to be enabled in your account.

Reviewer 5: @mariana-caetano This for example is no longer true, and users can self-install via the FastStore WebOps onboarding process.

> - If the **Content** option is not available, [open a ticket with VTEX Support](https://help.vtex.com/en/support) requesting the CMS to be enabled in your account.

Author: ```suggestion
```

> > ⚠️ The CMS is currently available for stores running [FastStore v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version) or higher. If your store uses the [Headless CMS (legacy)](https://developers.vtex.com/docs/guides/faststore/headless-cms-overview), see the [Upgrading from Headless CMS (legacy) to CMS](https://developers.vtex.com/docs/guides/upgrading-from-headless-cms-legacy-to-cms-overview) track first.

Reviewer 5: @mariana-caetano This is not accurate. CMS is currently compatible with FastStore v3 and v4.

> > ⚠️ The CMS is currently available for stores running [FastStore v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version) or higher. If your store uses the [Headless CMS (legacy)](https://developers.vtex.com/docs/guides/faststore/headless-cms-overview), see the [Upgrading from Headless CMS (legacy) to CMS](https://developers.vtex.com/docs/guides/upgrading-from-headless-cms-legacy-to-cms-overview) track first.

Author: ```suggestion
> ⚠️ The CMS is currently available for stores running Faststore v3 and [FastStore v4](https://developers.vtex.com/docs/guides/faststore/developer-tools-updating-the-cli-package-version). If your store uses the [Headless CMS (legacy)](https://developers.vtex.com/docs/guides/faststore/headless-cms-overview), see the [Upgrading from Headless CMS (legacy) to CMS](https://developers.vtex.com/docs/guides/upgrading-from-headless-cms-legacy-to-cms-overview) track first.
```

   ```jsonc  
   {  
     "$extends": [  
       "#/$defs/base-component"  
     ],  
     "$componentKey": "CallToAction",  
     "$componentTitle": "Call To Action",  
     "title": "Call To Action",  
     "description": "Get your 20% off on the first purchase!",  
     "type": "object",  
     "required": [  
       "title",  
       "link"  
     ],  
     "properties": {  
       "title": {  
         "title": "Title",  
         "type": "string"  
       },  
       "link": {  
         "title": "Link Path",  
         "type": "object",  
         "required": [  
           "text",  
           "url"  
         ],  
         "properties": {  
           "text": {  
             "title": "Text",  
             "type": "string"  
           },  
           "url": {  
             "title": "URL",  
             "type": "string"  
           }  
         }  
       }  
     }  
   }  
   ```

7. Generate the final schema to add the new section. To do this, run the following command in a terminal:

    ```bash  
    vtex content generate-schema cms/faststore/components cms/faststore/pages -o cms/faststore/schema.json -b vtex.faststore  
    ```

> 8. Open the `schema.json` file and check if the section was added to the section list. It will probably be like that:

Reviewer 4: ```suggestion
8. Open the `schema.json` file and check if the section was added to the section list. It'll probably be like that:
```

    ```json  
    …  
    {  
        "$ref": "#/components/CallToAction"  
    }  
    …  
    ```

### Step 2 - Verify the component locally

Before syncing your changes with the CMS, run the storefront locally to confirm your new component compiles without errors. This catches typos, missing imports, and invalid schemas before involving the CMS.

1. In the project root, start the FastStore development server:

    ```bash
    yarn dev
    ```

2. Open the development server in your browser and check the terminal output for compilation errors related to `CallToAction`.

> > ℹ️ The section will not render visually yet because it has not been added to any page in the CMS, which is done in [Step 4](#step-4---add-the-component-to-the-cms). At this point, you are only verifying that the dev server starts and that your component code, schema, and registration are valid.

Reviewer 4: ```suggestion
> ℹ️ The section will not render visually yet because it has not been added to any page in the CMS, which is done in [Step 4](#step-4---add-the-component-to-the-cms). At this point, you're only verifying that the development server starts and that your component code, schema, and registration are valid.
```

> If the dev server starts successfully and reports no compilation errors, you are ready to sync your changes with the CMS.

Reviewer 4: ```suggestion
If the development server starts successfully and reports no compilation errors, you're ready to sync your changes with the CMS.
```

### Step 3 - Sync the changes with the CMS

> Push the changes to sync with the CMS. Before doing it, make sure you are logged in to your VTEX account by running in the terminal the `vtex login {accountName}` command. Change the `{accountName}` to your store account, for example, `vtex login mystore`.

Reviewer 4: ```suggestion
Push the changes to sync with the CMS. Before doing it, make sure you're logged in to your VTEX account by running in the terminal the `vtex login {accountName}` command. Change the `{accountName}` to your store account, for example, `vtex login mystore`.
```

Push the schema generated to the CMS to reflect in the Admin. To do this, run the following command:

```shell
vtex content upload-schema cms/faststore/schema.json
```

### Step 4 - Add the component to the CMS

> 1. In the Admin, open **Storefront > Content**, and click a Content-Type, such as **Home**, to check whether the component was added.

Reviewer 5: @mariana-caetano Please update the language from "Content-Type" to "Entry". In this example you're clicking on an Entry.

> 1. In the Admin, open **Storefront > Content**, and click a Content-Type, such as **Home**, to check whether the component was added.

Author: ```suggestion
1. In the Admin, open **Storefront > Content**, and click an Entry, such as **Home**, to check whether the component was added.
```
